src/io/config.cpp

```diff
@@ -356,14 +356,24 @@ void Config::CheckParamConflict() {
                  tree_learner.c_str());
     }
   }
-  // Check max_depth and num_leaves
-  if (max_depth > 0) {
+
+  // max_depth defaults to -1, so max_depth>0 implies "you explicitly overrode the default"
+  //
+  // Changing max_depth while leaving num_leaves at its default (31) can lead to 2 undesirable situations:
+  //
+  //   * (0 <= max_depth <= 4) it's not possible to produce a tree with 31 leaves
```

Comment:
Do we also want to emit a warning for this case actually?

Reply:
I don't think we should.

In this case, LightGBM is going to try to fit a full depth-wise tree.

So all these combinations would be the same:

```text
max_depth=4, (omitted num_leaves)
max_depth=4, num_leaves=31
max_depth=4, num_leaves=21
max_depth=4, num_leaves=16
```

So yeah, if you don't understand that you might end up wasting some cross-validation runs on parameter combinations that will generate the same model, but you won't get a worse fit.

I think it's the right balance to generally defer figuring those things out to users' evaluation setups.

I think the *new* warning added in this PR is a worthwhile addition because it tells users about a situation that is non-obvious and which could *substantially* affect performance. e.g. if you set `max_depth=8` and omit `num_leaves`, you have 256 leaves available to you and are only using up to 31 of them... that's a pretty big gap, that isn't obvious.

Reply:
All right, fair enough! :+1: